File: why_LeetCode/Day03/C++/Longest Consecutive Sequence.cpp
#方法一
'''
   如果允许O(n log n)的复杂度，那么可以先排序，可是本题要求O(n)。
   由于序列里的元素是无序的，又要求O(n)，首先要想到用哈希表。
   用一个哈希表unordered_ map<int, bool> used 记录每个元素是否使用,对每个元素，以该元
   素为中心，往左右扩张，直到不连续为止，记录下最长的长度。
'''

class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_map<int,bool> used;
        for(auto i : nums) used[i]=false;
        int longest=0;

        for(auto i : nums)
        {
            if (used[i]) continue;

            int length=1;
            used[i]=true;
            for(int j=i+1;used.find(j)!=used.end();j++)
            {
                used[j]=true;
                length++;
            }
            for(int j=i-1;used.find(j)!=used.end();j--)
            {
                used[j]=true;
                length++;
            }
            longest=max(longest,length);
        }
        return longest;
    }
};

#方法二：
'''
第一直觉是个聚类的操作，应该有union,find 的操作.连续序列可以用两端和长度来表;
示.本来用两端就可以表示， 但考虑到查询的需求，将两端分别暴露出来.用unordered_-
map<int，int> map 来存储.

'''
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        if(nums.size()==0)
            return 0;
        
        unordered_map<int,int> map;
        int length=1;
        for(int i=0;i<nums.size();i++)
        {
            if(map.find(nums[i])!=map.end()) continue;

            map[nums[i]]=1;
            if(map.find(nums[i]-1)!=map.end())
            {
                length=max(length,mergeCluster(map,nums[i]-1,nums[i]));
            }
            if(map.find(nums[i]+1)!=map.end())
            {
                length=max(length,mergeCluster(map,nums[i],nums[i]+1));
            }   
        }
        return length;

    }

private:
    int mergeCluster(unordered_map<int, int> &map, int left, int right)
    {
        int upper=right+(map[right]-1);
        int lower=left-(map[left]-1);
        int length=upper-lower+1;
        map[upper]=length;
        map[lower]=length;
        return length;
    }
};

File: why_LeetCode/Day03/Python/Longest Consecutive Sequence.py
'''
类似于c++的第二种方法。
建立一个哈希表，key是区间端点，val是这段区间的长度。


线性扫描输入数组，如果当前元素已经在哈希表里，则跳过，


如果不在哈希表里，就看一下它左右两侧区间的长度left，right，
计算出它自身的区间长度length = 1 + left + right。


计算完之后更新新的左右端点的长度为，
len_dict[num - left] =length = len_dict[num + right],
len_dict[num]也赋值，为了占个位子。

重点：两个端点，以及区间长度。

'''


class Solution:
    def longestConsecutive(self, nums: List[int]) -> int:
        len_dict={}
        longest=0
        
        for num in nums:
            if num not in len_dict:
                left=len_dict.get(num-1,0)
                right=len_dict.get(num+1,0)
                
                length=left+right+1;
                longest=max(longest,length)
                
                for i in [num-left,num,num+right]:
                    len_dict[i]=length
        return longest